#11825: Generate new FormKey and replace for oldRequestParams Wishlist

osrecio opened the pull request:
Generate new FormKey afterLogin and set to the beforeRequest(Wishlist)

### Description
Generate new FormKey afterLogin and set to the beforeRequest(Wishlist)

### Fixed Issues (if relevant)
1. magento/magento2#11825: 2.1.9 Item not added to the Wishlist if the user is not logged at the moment he click on the button to add it
2. magento/magento2#11908: Adding to wishlist doesn't work when not logged in

### Manual testing scenarios
1. Create an user account.
2. Logout from the user account
3. Add a product to your Wishlist , you will get redirected to the login page
4. Login

### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

osrecio committed:
Generate new FormKey and replace for oldRequestParams

osrecio commented:
Can you check @vkublytskyi . Thanks!

ihor-sviziev commented on app/code/Magento/PageCache/Observer/FlushFormKey.php:
```diff
@@ -22,13 +23,20 @@ class FlushFormKey implements ObserverInterface
+     * @param Session $customerSession
      */
-    public function __construct(CookieFormKey $cookieFormKey, DataFormKey $dataFormKey)
+    public function __construct(CookieFormKey $cookieFormKey, DataFormKey $dataFormKey, Session $customerSession)
```
This is not backward compatible change. Could you add $customerSession as optional parameter with fallback to retrieving this object via ObjectManger?

osrecio replied:
Thanks @ihor-sviziev I will work after mm17es in this fix.

ihor-sviziev commented on app/code/Magento/PageCache/Observer/FlushFormKey.php:
```diff
@@ -38,7 +46,14 @@ public function __construct(CookieFormKey $cookieFormKey, DataFormKey $dataFormK
         $this->dataFormKey->set(null);
+
+        $beforeParams = $this->customerSession->getBeforeRequestParams();
+        if ($beforeParams['form_key'] == $currentFormKey) {
```
Could you add check that $beforeParams['form_key'] is exist? $beforeParams might be empty array.
Also would be great to use strict equal === comparison instead of ==.

osrecio replied:
Thanks @ihor-sviziev I will work after mm17es in this fix.

osrecio commented:
@ihor-sviziev Changes Made. Thanks for your comments

vkublytskyi commented:
@ihor-sviziev *Observers* and *plugins* do not follow this backward compatibility rule (as they must not be extended or called individually) and constructor may be changed.

vkublytskyi commented:
@osrecio tests failed as this fix introduces dependency for PageCache on Customer module. We should avoid this for bugfixes. Also we should avoid this as Customer module already has a dependency on PageCashe module and we must avoid circular dependencies. As a solution, I may suggest make a plugin on `\Magento\PageCache\Observer\FlushFormKey` in a customer module

ihor-sviziev reviewed (changes requested):
Please review latest comments form @vkublytskyi

osrecio commented:
Hi @vkublytskyi I added a Plugin in Customer Module to make the same logic.

ihor-sviziev commented:
@osrecio static test failed because of not used parameter. Please ignore it in the code. 
Also would be great to remove not related changes to keep only needed.
And third thing - could you add unit tests or better integration tests for your new code?

vkublytskyi commented on app/code/Magento/Customer/Model/Plugin/CustomerFlushFormKey.php:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * @param FlushFormKey $subject
+     * @param callable $proceed
+     * @param $args
```
Please add `@SuppressWarnings(PHPMD.UnusedFormalParameter)` to avoid static tests failure

vkublytskyi commented on app/code/Magento/Customer/Model/Plugin/CustomerFlushFormKey.php:
```diff
@@ -0,0 +1,52 @@
+<?php
+/**
```
Please set copyright same as in other files for `2.1-develop`:

```php
/**
 * Copyright © 2013-2017 Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

osrecio committed:
Add Test CustomerFlushFormKey Plugin

osrecio commented:
@vkublytskyi @ihor-sviziev 
1. Added `@SuppressWarnings(PHPMD.UnusedFormalParameter)` to avoid static tests failure
2. Deleted not related changes
3. Added Unit Tests

vkublytskyi commented on app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerFlushFormKeyTest.php:
```diff
@@ -0,0 +1,110 @@
+            ->method('setBeforeRequestParams')
+            ->with($beforeParams);
+
+        $plugin->aroundExecute($observer, $this->closure, $observer);
```
This line contains an error. `$observer` is instance of `Magento\PageCache\Observer\FlushFormKey` which is child of `\Magento\Framework\Event\ObserverInterface`. `aroundExecute` method should mutch signature of `Magento\PageCache\Observer\FlushFormKey::execute` so last parameter should be instance of `Magento\Framework\Event\Observer` which is child of `Magento\Framework\DataObject` and NOT related to `Magento\Framework\Event\ObserverInterface`.

To catch such kind of an issue I propose you to add logic to closure and invoke `FlushFormKey ::execute` inside.

vkublytskyi reviewed (changes requested):
Added unit test contains an error

osrecio commented:
Thanks @vkublytskyi I was going to work on the changes this weekend. Busy Week

Commit:
MAGETWO-83287: #11825: Generate new FormKey and replace for oldRequestParams Wishlist #12038

 - fixed unit test

vkublytskyi commented:
@osrecio I will resolve minor fixes to free your time for more interesting things)
